Move MicrotaskQueue to JSC

Constellation opened the pull request:
#### 57d6328a4f3a6643f787123035739b461c9e0593
<pre>
Move MicrotaskQueue to JSC
<a href="https://bugs.webkit.org/show_bug.cgi?id=289131">https://bugs.webkit.org/show_bug.cgi?id=289131</a>
<a href="https://rdar.apple.com/146170668">rdar://146170668</a>

Reviewed by Ryosuke Niwa.

This patch unifies JSC&apos;s MicrotaskQueue and WebCore&apos;s MicrotaskQueue.
Integrating WebCore only concepts into JSC&apos;s MicrotaskQueue so that
WebCore can directly use JSC&apos;s MicrotaskQueue.

* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::queueMicrotask):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):
(): Deleted.
(JSC::createJSMicrotask): Deleted.
(JSC::JSMicrotask::run): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/Microtask.h:
(JSC::Microtask::Microtask): Deleted.
(JSC::Microtask::~Microtask): Deleted.
(JSC::Microtask::identifier const): Deleted.
* Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
(JSC::QueuedTask::tryRun):
(JSC::QueuedTask::isRunnable const):
(JSC::MicrotaskQueue::~MicrotaskQueue):
(JSC::MicrotaskQueue::visitAggregateImpl):
(JSC::MicrotaskQueue::enqueue):
(JSC::MicrotaskQueue::performMicrotaskCheckpoint):
(JSC::MarkedMicrotaskDeque::hasMicrotasksForFullyActiveDocument const):
(JSC::MarkedMicrotaskDeque::visitAggregateImpl):
(JSC::QueuedTask::run): Deleted.
* Source/JavaScriptCore/runtime/MicrotaskQueue.h:
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::setDispatcher):
(JSC::QueuedTask::globalObject const):
(JSC::QueuedTask::job const):
(JSC::QueuedTask::arguments const):
(JSC::MarkedMicrotaskDeque::swap):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::callOnEachMicrotaskTick):
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp: Removed.
* Source/WebCore/bindings/js/JSDOMMicrotask.h: Removed.
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::JSDOMWindow::queueMicrotask):
* Source/WebCore/bindings/js/JSExecState.h:
(WebCore::JSExecState::runTask):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeCustom.cpp:
(WebCore::JSWorkerGlobalScope::queueMicrotask):
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::queueMicrotask):
(WebCore::EventLoop::run):
(WebCore::EventLoopTaskGroup::EventLoopTaskGroup):
(WebCore::EventLoopTaskGroup::~EventLoopTaskGroup):
(WebCore::EventLoopTaskGroup::jsMicrotaskDispatcher const):
(WebCore::EventLoopTaskGroup::queueMicrotask):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::WebCoreMicrotaskDispatcher::currentRunnability const):
(WebCore::MicrotaskQueue::MicrotaskQueue):
(WebCore::MicrotaskQueue::append):
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
(WebCore::MicrotaskQueue::hasMicrotasksForFullyActiveDocument const):
* Source/WebCore/dom/Microtasks.h:
(WebCore::WebCoreMicrotaskDispatcher::WebCoreMicrotaskDispatcher):
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::addExtendLifetimePromise):
(): Deleted.

Canonical link: <a href="https://commits.webkit.org/291649@main">https://commits.webkit.org/291649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb0890cc30bfe773a26b8d17f390021d58d82497

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98562 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44085 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95609 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13418 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21575 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10040 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51811 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9722 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43399 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86269 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100594 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92225 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20611 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80489 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20863 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80543 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79823 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24360 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13759 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15001 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20595 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/25773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114875 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23742 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.h:
```diff
@@ -33,37 +33,65 @@
 
-    QueuedTask(MicrotaskIdentifier identifier, JSValue job, JSValue argument0, JSValue argument1, JSValue argument2, JSValue argument3)
-        : m_identifier(identifier)
+    enum class Status : uint8_t {
```
Hm... how about we call this `Result`? `Status` sounds like something that describes the current status of a task. Here, this enum is more like a result of trying to execute a task.

Constellation replied:
That sounds like a good idea. Changed.

rniwa commented on Source/WebCore/dom/EventLoop.cpp:
```diff
@@ -488,11 +545,68 @@ void EventLoopTaskGroup::queueTask(TaskSource source, EventLoop::TaskFunction&&
+{
+    queueMicrotask(JSC::QueuedTask {
+        EventLoopFunctionMicrotaskDispatcher::create(*this, WTFMove(function)),
+        JSC::MicrotaskIdentifier::generate(),
```
Can we add a `QueuedTask` constructor which takes two arguments instead of setting all these arguments to `nullptr` or `{ }`?

Constellation replied:
Sounds good. Changed.

rniwa commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -258,16 +301,13 @@ void JSDOMWindowBase::queueMicrotaskToEventLoop(JSGlobalObject& object, Ref<JSC:
+    if (!userGestureToken)
+        task.setDispatcher(eventLoop.jsMicrotaskDispatcher());
+    else
+        task.setDispatcher(UserGestureInitiatedMicrotaskDispatcher::create(eventLoop, Ref { *userGestureToken }));
```
Hm... it seems a bit wasteful to create a new object for every user gesture initiated task but it's no worse than the old code.

Constellation replied:
Yeah, I wonder if we can have some optimizations (like, consider introducing token into QueuedTask), but for now, keeping it as the same level efficiency as the old code.

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
```diff
@@ -37,22 +41,106 @@ namespace JSC {
+        case QueuedTask::Status::Executed: {
+            break;
+        }
+        case QueuedTask::Status::Stopped: {
```
Maybe we should call this value `Discard` instead?

Constellation replied:
Sounds good. Removed.

rniwa commented on Source/WebCore/dom/EventLoop.cpp:
```diff
@@ -406,6 +406,63 @@ Markable<MonotonicTime> EventLoop::nextTimerFireTime() const
+
+    JSC::QueuedTask::Status run(JSC::QueuedTask& task) final
+    {
+        auto group = m_group.get();
```
Hm... this code is repeated like 3 different times. Can we share the code somehow? Maybe we can have something like `WebCoreMicrotaskDispatcher` which knows what to return as a result?

Constellation replied:
Yeah, we can have `WebCoreMicrotaskDispatcher` and having `EventLoopTaskGroup` weak pointer in it. Sounds good. Changed.

rniwa commented on Source/WebCore/dom/Microtasks.cpp:
```diff
@@ -61,25 +62,8 @@ void MicrotaskQueue::performMicrotaskCheckpoint()
-    m_microtaskQueue = WTFMove(toKeep);
 
     if (!vm->executionForbidden()) {
         auto checkpointTasks = std::exchange(m_checkpointTasks, { });
```
Looks like this whole facility exists just for indexedDB's `IDBTransaction::deactivate()`. I wonder if it can even run JavaScript in it...

rniwa replied:
Sihui tells me this can dispatch events so indeed we can end up running arbitrary scripts.

Constellation replied:
I see. Eventually we could make them abstracted and move them to JSC side. For now, it is keeping this as WebCore specific operations (IDB).

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
```diff
@@ -37,22 +41,106 @@ namespace JSC {
+        }
+
+        switch (status) {
+        case QueuedTask::Status::Executed: {
```
These curly braces for case statements here seem unnecessary?

Constellation replied:
Yes, removed.

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
```diff
@@ -37,22 +41,106 @@ namespace JSC {
+{
+    for (auto& task : m_queue) {
+        if (task.isRunnable())
+            return true;
```
It seems a bit inefficient to traverse through all these tasks to find the runnable one although in most cases, all tasks are runnable so we should exit in a couple of iterations. It's probably okay for this PR though.

Constellation replied:
Sounds good. Will do it in a follow-up PR.

rniwa commented on Source/WebCore/dom/Microtasks.h:
```diff
@@ -22,12 +22,15 @@
 #pragma once
 
 #include "EventLoop.h"
+#include <JavaScriptCore/MicrotaskQueue.h>
```
If we're including this header, then we don't need a forward declaration of MicrotaskQueue down below, right?

Constellation replied:
Yes. Removed.

rniwa commented on Source/WebCore/dom/Microtasks.h:
```diff
@@ -39,22 +42,22 @@ class MicrotaskQueue final {
     // For the main thread the VM lives forever. For workers it's lifetime is tied to our owning WorkerGlobalScope. Regardless, we retain the VM here to be safe.
     Ref<JSC::VM> m_vm;
     WeakPtr<EventLoop> m_eventLoop;
+    std::unique_ptr<JSC::MicrotaskQueue> m_microtaskQueue;
```
Maybe this could be just a plain old member variable?

Constellation replied:
Ah! Yup, right. Changed.

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.h:
```diff
@@ -33,37 +33,65 @@
     MicrotaskIdentifier m_identifier;
-    JSValue m_job;
-    JSValue m_arguments[maxArguments];
+    JSGlobalObject* m_globalObject { nullptr };
```
We should figure out a way to avoid having all these raw pointer member variables in JSC code at some point.

Constellation replied:
For now, I think this is fine as this is designed to be particularly tied to GC itself (MicrotaskQueue is coordinating GC scanning specially etc.), so yeah, this is really special code in terms of GC.

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
```diff
@@ -37,22 +41,106 @@ namespace JSC {
+QueuedTask::Status QueuedTask::tryRun()
 {
+    if (m_dispatcher)
+        return m_dispatcher->run(*this);
```
Per safer C++ guidelines, this should be:
```cpp
if (RefPtr dispatcher = m_dispatcher)
    dispatcher->run(*this)
```

Constellation replied:
Changed.

rniwa commented on Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
```diff
@@ -37,22 +41,106 @@ namespace JSC {
+{
+    if (!m_dispatcher)
+        return true;
+    return m_dispatcher->isRunnable();
```
We should store `m_dispatcher` in a local `RefPtr` before calling `isRunnable()`.

Constellation replied:
Changed.

Constellation reviewed (commented):
Thanks!

Jarred-Sumner commented on Source/JavaScriptCore/runtime/MicrotaskQueue.h:
```diff
@@ -33,37 +33,73 @@
+    std::span<const JSValue> arguments() const { return std::span { m_arguments }; }
 
 private:
+    RefPtr<MicrotaskDispatcher> m_dispatcher;
```
Is there likely going to be many different dispatcher pointers per JSC::VM? I wonder if this could be passed along to the function that runs the microtasks

Constellation replied:
WebCore can create a new dispatcher arbitrarily per task, so we need some mechanism to dispatch some special function like this dispatcher.

Constellation committed:
Move MicrotaskQueue to JSC
https://bugs.webkit.org/show_bug.cgi?id=289131
rdar://146170668

Reviewed by Ryosuke Niwa.

This patch unifies JSC's MicrotaskQueue and WebCore's MicrotaskQueue.
Integrating WebCore only concepts into JSC's MicrotaskQueue so that
WebCore can directly use JSC's MicrotaskQueue.

* Source/JavaScriptCore/runtime/GlobalObjectMethodTable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::queueMicrotask):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runJSMicrotask):
(): Deleted.
(JSC::createJSMicrotask): Deleted.
(JSC::JSMicrotask::run): Deleted.
* Source/JavaScriptCore/runtime/JSMicrotask.h:
* Source/JavaScriptCore/runtime/Microtask.h:
(JSC::Microtask::Microtask): Deleted.
(JSC::Microtask::~Microtask): Deleted.
(JSC::Microtask::identifier const): Deleted.
* Source/JavaScriptCore/runtime/MicrotaskQueue.cpp:
(JSC::QueuedTask::tryRun):
(JSC::QueuedTask::isRunnable const):
(JSC::MicrotaskQueue::~MicrotaskQueue):
(JSC::MicrotaskQueue::visitAggregateImpl):
(JSC::MicrotaskQueue::enqueue):
(JSC::MicrotaskQueue::performMicrotaskCheckpoint):
(JSC::MarkedMicrotaskDeque::hasMicrotasksForFullyActiveDocument const):
(JSC::MarkedMicrotaskDeque::visitAggregateImpl):
(JSC::QueuedTask::run): Deleted.
* Source/JavaScriptCore/runtime/MicrotaskQueue.h:
(JSC::QueuedTask::QueuedTask):
(JSC::QueuedTask::setDispatcher):
(JSC::QueuedTask::globalObject const):
(JSC::QueuedTask::job const):
(JSC::QueuedTask::arguments const):
(JSC::MarkedMicrotaskDeque::swap):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::~VM):
(JSC::VM::drainMicrotasks):
(JSC::VM::beginMarking):
(JSC::VM::visitAggregateImpl):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::callOnEachMicrotaskTick):
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp: Removed.
* Source/WebCore/bindings/js/JSDOMMicrotask.h: Removed.
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::JSDOMWindow::queueMicrotask):
* Source/WebCore/bindings/js/JSExecState.h:
(WebCore::JSExecState::runTask):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.cpp:
(WebCore::JSShadowRealmGlobalScopeBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.cpp:
(WebCore::JSWorkerGlobalScopeBase::queueMicrotaskToEventLoop):
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeCustom.cpp:
(WebCore::JSWorkerGlobalScope::queueMicrotask):
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::queueMicrotask):
(WebCore::EventLoop::run):
(WebCore::EventLoopTaskGroup::EventLoopTaskGroup):
(WebCore::EventLoopTaskGroup::~EventLoopTaskGroup):
(WebCore::EventLoopTaskGroup::jsMicrotaskDispatcher const):
(WebCore::EventLoopTaskGroup::queueMicrotask):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::WebCoreMicrotaskDispatcher::currentRunnability const):
(WebCore::MicrotaskQueue::MicrotaskQueue):
(WebCore::MicrotaskQueue::append):
(WebCore::MicrotaskQueue::performMicrotaskCheckpoint):
(WebCore::MicrotaskQueue::hasMicrotasksForFullyActiveDocument const):
* Source/WebCore/dom/Microtasks.h:
(WebCore::WebCoreMicrotaskDispatcher::WebCoreMicrotaskDispatcher):
* Source/WebCore/workers/service/ExtendableEvent.cpp:
(WebCore::ExtendableEvent::addExtendLifetimePromise):
(): Deleted.

Canonical link: https://commits.webkit.org/291649@main
